### ISSUE TITLE
ConfigProxy crashes with EMFILE error

### ISSUE THREAD

**Issue by kafonek** [jupyterhub/jupyterhub#900]
I am seeing my ConfigProxy crash and not recover after upgrading to Jupyterhub 0.7.0 and configurable-http-proxy 1.3.1.  Unfortunately I cannot copy/paste the entire output log because the logs are on a separate network, but here is the general trace-back:

JupyterHub is running behind an nginx which is handling authentication and reverse proxying, I have a custom authenticator class that is just using headers that nginx inserted.

Logs start off looking okay:
```
Hub API listening on http://127.0.0.1:8081/jupyter/hub
Starting Proxy @ http://*:8002/jupyter
[ConfigProxy] Proxying http://*:8002 to http:127.0.0.1:8081
[ConfigProxy] Proxy API at http://127.0.0.1:8003/api/routes
JupyterHub is now running at http://127.0.0.1:8002/jupyter
```

After that, I can open a few notebooks, navigate around directories on the hub, but before too long I run into a series of errors that look like this - 

```
[ConfigProxy] Proxy error: Error: connect EMFILE
[ConfigProxy] 503 GET /jupyter/user/kafonek/static/tree/js/main.min.js.map
[ConfigProxy] Failed to get custom error page Error: connect EMFILE
[ConfigProxy] Uncaught Exception TypeError: Object #<ClientRequest> has no method '_handle_proxy_error_default'
  at ClientRequest.<anonymous> (/usr/lib/node_modules/configurable-http-proxy/lib/configproxy.js:334:18)
  at ClientRequest.emit (events.js:95:17)
  at Socket.socketCloseListener (http.js:1527:9)
  at Socket.emit (events.js:117:20)
  at TCP.close (net.js:466:12)
```

Apologies again for not having the full log, hopefully someone has run into the same error and can copy/paste the full trace-back.  

Thanks,

**Comment by minrk** [comment]
@kafonek what version of nodejs (`node --version`)? The uncaught exception is in the code to show you a custom error page, and fixed by https://github.com/jupyterhub/configurable-http-proxy/pull/91, but the original error is file descriptor exhaustion, and appears to be sensitive to the underlying version of nodejs. If you are using nodejs < 4, you might try upgrading.

**Comment by kafonek** [comment]
@minrk, node --version gives me v0.10.36.

We seem to be able to reproduce the error by just attempting to open many different notebooks at the same time.  Is there a work-around I can put in temporarily that will identify when configurable-http-proxy has hit this Uncaught exception and restart the proxy?

Thanks for all of your work!

**Comment by kafonek** [comment]
@minrk, after upgrading to node 6.4.4, we have been so far unable to crash our proxy even when we open up many notebooks and files at once, which would always crash the proxy before.  Thanks for all of your work!

If you want to close this ticket with a note that upgrading node seems to fix the problem, that would be fine by me.  Leaving it open for now in case you wanted to add anything into configurable-http-proxy for old node versions.

**Comment by minrk** [comment]
Yup, 0.10 is still what ships with some common Linux distros, and the issue appears to be fixed in node 0.11. It has to do with node-http-proxy not reusing connections, so it can run out of file descriptors pretty easily under load. configurable-http-proxy 2.0 will not support node < 4, FWIW.

cf https://github.com/nodejitsu/node-http-proxy/issues/570

**Comment by willingc** [comment]
Thanks @kafonek @minrk. Closing with the following resolution.

---
## Resolution

Some Linux system install still use an old version of node (such as 0.10). Upgrading the node package to a more recent version such as 6.4.4 will resolve the issue of the crashing proxy. As a note, configurable-http-proxy 2.0 will not support node < 4.
